### ISSUE TITLE
oclif promote shows NoSuchKey: The specified key does not exist.

### ISSUE THREAD

**Issue by tlkiong** [oclif/oclif#770]
I am not sure if this is a bug. However, based on the docs [here](https://oclif.io/docs/releasing#standalone-tarballs), I should be able to push to S3 by using the command `oclif promote`.

However, after some fiddling around and trying to understand what is going wrong due to lack of documentation, I encounter this issue:

```
    NoSuchKey: The specified key does not exist.
    Code: NoSuchKey
```

**Comment by RodEsp** [comment]
Hey @tlkiong, if you're using Oclif v2 you'll need to use `oclif upload` first to upload your CLI to S3 and then you can use `oclif promote` to promote the CLI to a specific release channel.

We are currently working on updating and improving the docs. Sorry for the hassle.
Let me know if that helps!

**Comment by creatzor** [comment]
I get this error because it attempts to copy `<app-name>-v<version>-<sha>-<target>-buildmanifest` however, the upload command never uploaded that file.

**Comment by mattbishop** [comment]
Same failure for me at the `buildmanifest` file. I would expect `oclif pack tarballs` to generate that file, but it does not.

I am following the directions here: https://oclif.io/docs/releasing

**Comment by mattbishop** [comment]
It looks like oclif expects the `*-buildmanifest` file to be in the S3 bucket already?

https://github.com/oclif/oclif/blob/db03bd0ddc1fedb5930bb21dd4aa75cff035e420/test/integration/publish.test.ts#L52

If so, how do I create this file? What are its contents? Is there an oclif command to generate and push? Feels like there is a missing step in the instructions for tarball releases.

**Comment by mattbishop** [comment]
Just realized this was a test file, but still the question holds--how does one create a `{arch}-buildmanifest` file?

**Comment by DaanSchoukens** [comment]
Wondering the same thing

**Comment by knvpk** [comment]
Facing same problem. is this addressed ?

**Comment by RodEsp** [comment]
Hey @mattbishop, @DaanSchoukens, @creatzor. Looking into this right now.

I think we're missing a check in the `upload` and `promote` commands that exists for `pack`.
Looks like if there's no `s3.host` in the oclif config the `-buildmanifest` files won't be built. See [here](https://github.com/oclif/oclif/blob/main/src/tarballs/build.ts#L120).

Do you have `oclif.update.s3.host` set in your `package.json` as described in the [Autoupdater docs](https://oclif.io/docs/releasing#autoupdater)?

**Comment by mattbishop** [comment]
Thank you @RodEsp I hadn't set that value and am using Backblaze. That property is required for the upload command to work. The `host` property actually needs to be a URL, like `https://s3.other-service.com` instead of just the host name.

Also, an undocumented requirement for non-AWS services is that the environment variable `AWS_S3_ENDPOINT` must be set to the host name, not a URL.

Thanks again for your help.

**Comment by RodEsp** [comment]
Thanks for the thorough follow up @mattbishop! I'll make sure we get that added to the docs soon.

**Comment by KASOGIT** [comment]
Hi @RodEsp , sorry to bother you so long after this issue have been written but i still have the problem trying to promote a macos build.

My config:

```
  "oclif": {
    "bin": "pos-cli",
    "dirname": "pos-cli",
    "commands": "./dist/commands",
    "macos": {
      "identifier": "com.hero.pos-cli",
      "host": "https://xxxx.s3.eu-west-3.amazonaws.com"
    },
    "update": {
      "s3": {
        "bucket": "prod-hero-post-transfer-cli-releases-2da7095475674",
        "host": "https://xxxx.s3.eu-west-3.amazonaws.com"
      }
    },
    "plugins": [
      "@oclif/plugin-help",
      "@oclif/plugin-plugins",
      "@oclif/plugin-update"
    ],
    "topicSeparator": " ",
    "topics": {
      "deploy": {
        "description": "deploy the new version of the app"
      },
      "package": {
        "description": "package the app"
      }
    }
  },
```

I'm running command in this order:
```
./node_modules/.bin/oclif pack macos
./node_modules/.bin/oclif upload macos
./node_modules/.bin/oclif promote --macos --sha=$(git rev-parse --short HEAD) --version=$(node -p "require('./package.json').version")
```

Getting:

```
Promoting buildmanifests & unversioned tarballs to stable
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-linux-x64-buildmanifest to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-linux-x64-buildmanifest
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-linux-x64.tar.gz to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-linux-x64.tar.gz
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4[16](https://github.com/Hero-payments/mono/actions/runs/4562908601/jobs/8050724915#step:9:17)5d/pos-cli-v0.0.0-eb4165d-linux-arm-buildmanifest to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-linux-arm-buildmanifest
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-linux-arm.tar.gz to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-linux-arm.tar.gz
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-win32-x64-buildmanifest to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-win32-x64-buildmanifest
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-win32-x64.tar.gz to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-win32-x64.tar.gz
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-win32-x86-buildmanifest to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-win32-x86-buildmanifest
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-win32-x86.tar.gz to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-win32-x86.tar.gz
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-darwin-x64-buildmanifest to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-darwin-x64-buildmanifest
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-darwin-x64.tar.gz to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-darwin-x64.tar.gz
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-darwin-arm64-buildmanifest to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-darwin-arm64-buildmanifest
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-darwin-arm64.tar.gz to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-darwin-arm64.tar.gz
Promoting macos pkgs to stable
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-x64.pkg to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-x64.pkg
oclif: s3:copyObject from s3://prod-hero-post-transfer-cli-releases-2da7095475674/versions/0.0.0/eb4165d/pos-cli-v0.0.0-eb4165d-arm64.pkg to s3://prod-hero-post-transfer-cli-releases-2da7095475674/channels/stable/pos-cli-arm64.pkg
    NoSuchKey: The specified key does not exist.
    Code: NoSuchKey
```

Any idea of what happening here and why no manifest are being generated during packing step ?

**Comment by KASOGIT** [comment]
> 

Yeap i tried everything listed above but i couldn't succeed to generate buildmanifest files. Do you have an working exemple for s3 that generate those files with the command `oclif pack macos` ? 

I always end up with *.pkg but no buildmanifest

<img width="1010" alt="image" src="https://user-images.githubusercontent.com/15908751/229122248-52605fab-bac9-4a7c-b854-4e74515478fc.png">

**Comment by RodEsp** [comment]
I don't have a working example unfortunately. You might ask @mdonnalley, he might be able to provide you with one.

**Comment by KASOGIT** [comment]
Thanks @RodEsp ! It would be awesome if you can provide me one @mdonnalley or if you already know what i have to do in order to generate those buildmanifest files

**Comment by KASOGIT** [comment]
Hi again @RodEsp, did u have the opportunity to take a look at my problem ?

**Comment by ryandagg** [comment]
We have a similar issue, and were able to track down which file is actually missing. This is not straightforward because all the copies appear to happen in parallel and the logging only tells you when the process starts.
This file does not exist in S3:
"oclif: s3:copyObject from s3://***/versions/8.0.4/96472c6/apt/heroku_8.0.4.96472c6-1_`i386`.deb to ..."

`oclif upload:deb` does upload files for `armel` such as "cli/dist/deb/heroku_8.0.3.ddf4a4b-1_`armel`.deb", but nothing for `i386`

I'm not sure how `oclif pack deb` decides targets, but it's not in line with `oclif upload deb`. I'm not sure if this is a bug or a configuration issue.

**Comment by ryandagg** [comment]
I spent a good chunk of time looking into this yesterday, and I can find no way to pack/create files ending in _i386. 
It looks like this line should be deleted, or caught to prevent the job from failing.
https://github.com/oclif/oclif/blob/main/src/commands/promote.ts#L155

**Comment by cristiand391** [comment]
Hey all, @ryandagg and I were looking at this failure in the CLI promote pipeline, here's what we found:

1) `oclif pack:deb` generates deb pkgs for the targets defined in `buildConfig.targets`:
https://github.com/oclif/oclif/blob/main/src/commands/pack/deb.ts#L94

these targets can be set by users in the CLI pjson (`pjson.oclif.update.node.targets`), and will default to these if the previous object is empty:
https://github.com/oclif/oclif/blob/main/src/tarballs/config.ts#L13-L20

So, `pack:deb` will not generate a deb pkg for `linux-x86`. But even if you set that target manually in your pjson, oclif will fail to build the pkg because node dropped 32bits build for linux a long time ago:
https://github.com/nodejs/build/issues/885

2) `oclif upload:deb` tries to upload the `i386` pkg:
https://github.com/oclif/oclif/blob/f483d1cc509a11fd9c255655cf29fc099be1f6f6/src/commands/upload/deb.ts#L54

but skips it here because the file doesn't exists:
https://github.com/oclif/oclif/blob/f483d1cc509a11fd9c255655cf29fc099be1f6f6/src/commands/upload/deb.ts#L47

3) `oclif promote:deb` picks targets the same way `pack:deb` does **BUT** you can also specify them via the `--targets` flag:
https://github.com/oclif/oclif/blob/f483d1cc509a11fd9c255655cf29fc099be1f6f6/src/commands/promote.ts#L31
https://github.com/oclif/oclif/blob/f483d1cc509a11fd9c255655cf29fc099be1f6f6/src/commands/promote.ts#L20

the odd thing is that when promoting the deb pkgs it just hardcodes the `i386` target (which again, hasn't worked for a very long time now)
https://github.com/oclif/oclif/blob/f483d1cc509a11fd9c255655cf29fc099be1f6f6/src/commands/promote.ts#L155-L156

Another issue with this is that `promote:deb` can never promote the `arm` deb pkg that `pack:deb` builds.

**Comment by tuler** [comment]
> 2. `oclif upload:deb` tries to upload the `i386` pkg:
>    https://github.com/oclif/oclif/blob/f483d1cc509a11fd9c255655cf29fc099be1f6f6/src/commands/upload/deb.ts#L54

My issue is that `oclif upload:deb` does not upload the `armel` pkg produced by `oclif pack:deb`.
Any reason why?

**Comment by cristiand391** [comment]
Sorry @tuler, can you confirm `oclif` builds a deb installer for `arm`? Ask because our CLI (Salesforce CLI) don't publish deb installers so we don't know the current state of it.

If so, seems it should be simple to fix that in `upload:deb`

**Comment by tuler** [comment]
I can confirm the arm package is produced by `oclif pack:deb`, it just is not published. I do think it's an easy fix, probably just add it [here](https://github.com/oclif/oclif/blob/4d4a96bbd65f65d124db5dbb7ee278109ea89238/src/commands/upload/deb.ts#L53).

**Comment by oap75** [comment]
> I can confirm the arm package is produced by `oclif pack:deb`, it just is not published. I do think it's an easy fix, probably just add it [here](https://github.com/oclif/oclif/blob/4d4a96bbd65f65d124db5dbb7ee278109ea89238/src/commands/upload/deb.ts#L53).

Any updates ?
